src/charm.py

```diff
@@ -276,55 +283,71 @@ def join_unit_to_cluster(self) -> None:
         Try to join the unit from the primary unit.
         """
         instance_label = self.unit.name.replace("/", "-")
-        instance_fqdn = self._get_unit_fqdn(self.unit.name)
+        instance_address = self._get_unit_fqdn(self.unit.name)
 
-        if self._mysql.is_instance_in_cluster(instance_label):
-            logger.debug("instance already in cluster")
-            return
-
-        # Add new instance to the cluster
-        try:
-            cluster_primary = self._get_primary_from_online_peer()
-            if not cluster_primary:
-                self.unit.status = WaitingStatus("waiting to get cluster primary from peers")
-                logger.debug("waiting: unable to retrieve the cluster primary from peers")
-                return
+        if not self._mysql.is_instance_in_cluster(instance_label):
+            # Add new instance to the cluster
+            try:
+                cluster_primary = self._get_primary_from_online_peer()
+                if not cluster_primary:
+                    self.unit.status = WaitingStatus("waiting to get cluster primary from peers")
+                    logger.debug("waiting: unable to retrieve the cluster primary from peers")
+                    return
+
+                if (
+                    self._mysql.get_cluster_node_count(from_instance=cluster_primary)
+                    == GR_MAX_MEMBERS
+                ):
+                    self.unit.status = WaitingStatus(
+                        f"Cluster reached max size of {GR_MAX_MEMBERS} units. Standby."
+                    )
+                    logger.warning(
+                        f"Cluster reached max size of {GR_MAX_MEMBERS} units. This unit will stay as standby."
+                    )
+                    return
+
+                # If instance is part of a replica cluster, locks are managed by the
+                # the primary cluster primary (i.e. cluster set global primary)
+                lock_instance = None
+                if self._mysql.is_cluster_replica(from_instance=cluster_primary):
+                    lock_instance = self._mysql.get_cluster_set_global_primary_address(
+                        connect_instance_address=cluster_primary
+                    )
 
-            if self._mysql.get_cluster_node_count(from_instance=cluster_primary) == GR_MAX_MEMBERS:
-                self.unit.status = WaitingStatus(
-                    f"Cluster reached max size of {GR_MAX_MEMBERS} units. Standby."
-                )
-                logger.warning(
-                    f"Cluster reached max size of {GR_MAX_MEMBERS} units. This unit will stay as standby."
+                # add random delay to mitigate collisions when multiple units are joining
+                # due the difference between the time we test for locks and acquire them
+                sleep(random.uniform(0, 1.5))
```

Comment:
why do we need this sleep? isnt getting a lock atomic in MySQL the way we acquire them?

Reply:
It's a hack to have less collisions on testing for the lock. I've noticed that, when triggered together, tries to acquire the lock takes longer to return then when the lock is committed with the random delay, rendering the process a bit snappier